### PR TITLE
Fix error reporting to produce call path in stack trace

### PR DESCRIPTION
**Pull request by redhug1**
### What

Update with code from dp-cantabular-filter-flex to show call path in stack trace

### How to review

Visual inspection.

Allready done locally by me to test, but if you wanted to:
You would have to run a full integration test on cantabular suite and change log lines to somthing like:
```
		return nil, errors.Wrapf(err, "for 'vaultPath': %s, failed in ReadKey", vaultPath)
```
and then cause the code to fail thus executing the above error.

What the code changes achieve is the following:
The existing stack trace produces:
```
"errors": [
  {
    "message": "GetCSVtoExcelStructure failed: error whilst getting CSV row StreamAndWrite failed, failed in getVaultKeyForCSVFile: for 'vaultPath': secret/shared/psk1/cantabular-example-1-49522c10-ac9f-4e06-849d-495d28023b16-2021-1.csv,  failed in ReadKey: key not found",
    "stack_trace": [
      {
        "file": "/go/.go/path/pkg/mod/github.com/!o!n!sdigital/log.go/v2@v2.0.9/log/log.go",
        "function": "github.com/ONSdigital/log.go/v2/log.Error",
        "line": 92
      },
      {
        "file": "/go/.go/path/pkg/mod/github.com/!o!n!sdigital/dp-kafka/v3@v3.1.1/consumer_handler.go",
        "function": "github.com/ONSdigital/dp-kafka/v3.(*ConsumerGroup).handleMessage",
        "line": 43
      },
      {
        "file": "/go/.go/path/pkg/mod/github.com/!o!n!sdigital/dp-kafka/v3@v3.1.1/consumer_handler.go",
        "function": "github.com/ONSdigital/dp-kafka/v3.(*ConsumerGroup).consumeMessage",
        "line": 81
      },
      {
        "file": "/usr/local/go/src/runtime/asm_amd64.s",
        "function": "runtime.goexit",
        "line": 1581
      }
    ]
  }
],
"event": "failed to handle message",
```

The new code produces:
```
"errors": [
  {
    "data": {
      "bucketName": "private-bucket",
      "commit_message": true,
      "err": {},
      "event": {
        "DatasetID": "cantabular-example-1-113b94c5-0bae-44cf-8b31-11b29ba0d02d",
        "Edition": "2021",
        "InstanceID": "461e0888-2210-418e-a7c4-2daa99a2588b",
        "RowCount": 19,
        "Version": "1"
      },
      "filenameCsv": "datasets/cantabular-example-1-113b94c5-0bae-44cf-8b31-11b29ba0d02d-2021-1.csv",
      "incomingCsvRow": 0
    },
    "message": "GetCSVtoExcelStructure failed: error whilst getting CSV row StreamAndWrite failed, failed in getVaultKeyForCSVFile: for 'vaultPath': secret/shared/psk1/cantabular-example-1-113b94c5-0bae-44cf-8b31-11b29ba0d02d-2021-1.csv, failed in ReadKey: key not found",
    "stack_trace": [
      {
        "file": "github.com/ONSdigital/dp-cantabular-xlsx-exporter/handler.(*XlsxCreate).getVaultKeyForCSVFile\n\t/go/handler/handler.go",
        "function": "(*XlsxCreate).getVaultKeyForCSVFile",
        "line": 120
      },
      {
        "file": "github.com/ONSdigital/dp-cantabular-xlsx-exporter/handler.(*XlsxCreate).StreamAndWrite\n\t/go/handler/handler.go",
        "function": "(*XlsxCreate).StreamAndWrite",
        "line": 81
      },
      {
        "file": "github.com/ONSdigital/dp-cantabular-xlsx-exporter/handler.(*XlsxCreate).GetCSVtoExcelStructure.func1\n\t/go/handler/handler.go",
        "function": "(*XlsxCreate).GetCSVtoExcelStructure.func1",
        "line": 294
      },
      {
        "file": "runtime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s",
        "function": "goexit",
        "line": 1581
      }
    ]
  }
],
"event": "failed to handle message",
```

Which is much better

### Who can review

Anyone

### CODE REVIEW

File: error.go [hunk 1]
```diff
@@ -80,3 +82,31 @@ func UnwrapLogData(err error) log.Data {
...
 	return logData
 }
+
+// stackTrace recursively unwraps the error looking for the deepest
+// level at which the error was wrapped with a stack trace from
+// github.com/pkg/errors (or conforms to the StackTracer interface)
+// and returns the slice of stack frames. These can are of type
+// log.go/EventStackTrace so can be used directly with log.Go's
+// available API to preserve the correct error logging format
+func stackTrace(err error) []log.EventStackTrace {
```
Comment (daiLlew) [inline error.go:92]: This looks fine to me - but its new functionality and therefore should have unit tests to prove it works as expected.

**Review by daiLlew — CHANGES_REQUESTED** [review]

**Review by DavidSubiros — COMMENTED** [review]
just a small comment, plus the issue pointed by @daiLlew , but the changes look good!
Also, the build failed, I think you need to add the package to go.mod?
```
error_interface.go:3:8: no required module provides package github.com/pkg/errors; to add it:
	go get github.com/pkg/errors
```

File: error.go [hunk 2]
```diff
@@ -80,3 +82,31 @@ func UnwrapLogData(err error) log.Data {
...
+	var serr stacktracer
+	var resp []log.EventStackTrace
+
+	for errors.Unwrap(err) != nil && errors.As(err, &serr) {
+		st := serr.StackTrace()
+		resp = make([]log.EventStackTrace, 0)
+		for _, f := range st {
+			line, _ := strconv.Atoi(fmt.Sprintf("%d", f))
```
[the remark this answers is not in the document]
Comment (ChrysmOre) [reply error.go:100]: I don't think we do here, as it's specifically for logging the stack trace we don't want to bail out or change any behaviour if an error is thrown, the only appropriate thing is to return a 0 (or a -1 might be slightly more clear to signify something's gone wrong but it's negligible)

File: consumer_group.go [hunk 3]
```diff
@@ -344,7 +344,12 @@ func (cg *ConsumerGroup) LogErrors(ctx context.Context) {
 				logData := UnwrapLogData(err)
 				logData["topic"] = cg.topic
 				logData["group_name"] = cg.group
-				log.Error(ctx, "received kafka consumer-group error", err, logData)
+				log.Info(ctx, "received kafka consumer-group error", &log.EventErrors{{
```
[the remark this answers is not in the document]
Comment (ChrysmOre) [reply consumer_group.go]: I think we can actually set it to ERROR if we use 
```
log.Info(ctx, "received kafka consumer-group error", log.ERROR, &log.EventErrors...)
```

**Review by ChrysmOre — CHANGES_REQUESTED** [review]

File: consumer_group.go [hunk 4]
```diff
@@ -344,7 +344,7 @@ func (cg *ConsumerGroup) LogErrors(ctx context.Context) {
 				logData := UnwrapLogData(err)
 				logData["topic"] = cg.topic
 				logData["group_name"] = cg.group
-				log.Warn(ctx, "received kafka consumer-group error", &log.EventErrors{{
+				log.Warn(ctx, "received kafka consumer-group error", log.ERROR, &log.EventErrors{{
```
[the remark this answers is not in the document]
Comment (redhug1) [reply consumer_group.go:347]: done

**Review by DavidSubiros — APPROVED** [review]
LGTM and I think everything has been addressed, but @daiLlew  should also approve once he's happy with it.

**Review by ChrysmOre — APPROVED** [review]
LGTM

**Review by daiLlew — APPROVED** [review]
LGTM. 

I do think that this could be possibly be split out into a generic library. The is really useful functionality that is not limited to just Kafka errors, by making a library any of our other apps could also benefit from this without having duplicate the functionality again.

Something to consider.

**Pull request merged**